### PR TITLE
Add inplace variants of cholesky and triangular solve operations

### PR DESCRIPTION
**Pull request by RSchwan**
## Description

This PR introduces inplace variants of `wp.tile_cholesky()`, `wp.tile_cholesky_solve()`, `wp.tile_lower_solve()` and `wp.tile_upper_solve()`. In a similar spirit to https://github.com/NVIDIA/warp/pull/1023, this can can reduce the used shared memory in a kernel.

## Before your PR is "Ready for review"

- [x] All commits are [signed-off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--s) to indicate that your contribution adheres to the [Developer Certificate of Origin](https://developercertificate.org/) requirements
- [x] Necessary tests have been added
- [x] Documentation is up-to-date
- [x] Auto-generated files modified by compiling Warp and building the documentation have been updated (e.g. `__init__.pyi`, `functions.rst`)
- [x] Code passes formatting and linting checks with `pre-commit run -a`

## Summary by CodeRabbit

* **New Features**
  * Added in-place tile operations for Cholesky, triangular solves, and FFT/IFFT to enable overwriting inputs and reduce memory use.

* **Documentation**
  * Public API docs updated to describe in-place semantics, supported datatypes, and clarified solve notation.

* **Tests**
  * New tests validating in-place Cholesky, forward/back substitution, and multi‑RHS scenarios.

* **Refactor**
  * Examples and kernels updated to use in-place variants, removing intermediate temporaries.

### CODE REVIEW

File: warp/tests/tile/test_tile_cholesky.py [hunk 1]
```diff
@@ -83,6 +83,42 @@ def test_tile_cholesky_cholesky(test, device):
...
+def test_tile_cholesky_cholesky_inplace(test, device):
+    rng = np.random.default_rng(42)
+    L_h = np.tril(rng.random((TILE_M, TILE_M)))  # Lower triangular matrix
+    A_h = L_h @ L_h.T
+    Y_h = np.arange(TILE_M, dtype=np.float64)
+
+    Y_sol_np = np.linalg.solve(A_h, Y_h)
+
+    A_wp = wp.array(A_h, requires_grad=True, dtype=wp.float64, device=device)
+    Y_wp = wp.array(Y_h, requires_grad=True, dtype=wp.float64, device=device)
```
Comment (daedalus5) [inline warp/tests/tile/test_tile_cholesky.py]: `requires_grad` isn't needed here

File: warp/tests/tile/test_tile_cholesky.py [hunk 2]
```diff
@@ -83,6 +83,42 @@ def test_tile_cholesky_cholesky(test, device):
...
+
+    wp.launch_tiled(tile_math_cholesky_inplace, dim=[1, 1], inputs=[A_wp, Y_wp], block_dim=TILE_DIM, device=device)
+    wp.synchronize_device(device)
```
Comment (daedalus5) [inline warp/tests/tile/test_tile_cholesky.py]: shouldn't need device synchronization

File: warp/tests/tile/test_tile_cholesky.py [hunk 3]
```diff
@@ -149,6 +185,58 @@ def test_tile_cholesky_cholesky_multiple_rhs(test, device):
...
+    rng = np.random.default_rng(42)
+    L_h = np.tril(rng.random((TILE_M, TILE_M)))  # Lower triangular matrix
+    A_h = L_h @ L_h.T
+    Y_h = np.arange(TILE_M * TILE_M, dtype=np.float64).reshape((TILE_M, TILE_M))
+    Z_h = np.zeros_like(Y_h)
+
+    Y_sol_np = np.linalg.solve(A_h, Y_h.T).T
+    Z_np = Y_sol_np @ Y_sol_np
+
+    A_wp = wp.array(A_h, requires_grad=True, dtype=wp.float64, device=device)
```
Comment (daedalus5) [inline warp/tests/tile/test_tile_cholesky.py]: Same situation here. No doubt this was modeled off the non-inplace test; in that case, the intention was that the backward pass may be implemented in the future. The inplace variants on the other hand won't have backward implementations, so it's safe to exclude gradient arrays.

File: warp/_src/builtins.py [hunk 4]
```diff
@@ -9313,7 +9313,7 @@ def tile_flip_layout(layout):
 ##
 def tile_fft_generic_value_func(arg_types, arg_values):
     if arg_types is None:
-        return tile(dtype=vector(length=2, dtype=Float), shape=Tuple[int, int])
+        return None
```
Comment (daedalus5) [inline warp/_src/builtins.py:9928]: Nice catch

File: warp/_src/builtins.py [hunk 5]
```diff
@@ -9563,9 +9589,37 @@ def tile_cholesky_generic_lto_dispatch_func(
...
+    value_func=tile_cholesky_inplace_generic_value_func,
+    lto_dispatch_func=tile_cholesky_inplace_generic_lto_dispatch_func,
+    variadic=True,
+    doc="""Compute the Cholesky factorization L of a matrix A.
+    L is lower triangular and satisfies LL^T = A.
+
+    Only the lower triangular portion of A is processed, and replaced by the lower triangular Cholesky factor L;
+    the upper triangular part is untouched.
+
+    Note that computing the adjoint is not yet supported.
```
Comment (daedalus5) [inline warp/_src/builtins.py]: You can remove this line

File: warp/_src/builtins.py [hunk 6]
```diff
@@ -9563,9 +9589,37 @@ def tile_cholesky_generic_lto_dispatch_func(
...
+    group="Tile Primitives",
+    export=False,
+    namespace="",
+    is_differentiable=False,
+)
+
+
+def _tile_cholesky_solve_generic_value_func(inplace: bool, arg_types, arg_values):
     if arg_types is None:
-        return None
```
Comment (daedalus5) [inline warp/_src/builtins.py:10180]: Nice catch

File: warp/_src/builtins.py [hunk 7]
```diff
@@ -9694,7 +9777,77 @@ def tile_cholesky_solve_generic_lto_dispatch_func(
...
-def tile_lower_solve_generic_lto_dispatch_func(
+add_builtin(
+    "tile_cholesky_solve_inplace",
+    input_types={"L": tile(dtype=Float, shape=Tuple[int, int]), "y": tile(dtype=Float, shape=Tuple[int])},
+    value_func=tile_cholesky_solve_inplace_generic_value_func,
+    lto_dispatch_func=tile_cholesky_solve_inplace_generic_lto_dispatch_func,
+    variadic=True,
+    doc="""With L such that LL^T = A, solve for x in Ax = y by overwriting y with x
+
+    Note that computing the adjoint is not yet supported.
```
Comment (daedalus5) [inline warp/_src/builtins.py]: Can be removed

File: warp/_src/builtins.py [hunk 8]
```diff
@@ -9694,7 +9777,77 @@ def tile_cholesky_solve_generic_lto_dispatch_func(
...
+    if y.shape[0] != l.shape[0]:
+        raise ValueError(
+            f"tile_lower_solve() 'y' argument must have the same number of elements as the number of rows in 'L', "
+            f"got {y.shape[0]} elements in 'y' and {l.shape[0]} rows in 'L'"
+        )
+
+    if inplace:
+        return None
+    return tile(dtype=l.dtype, shape=y.shape, layout=y.layout, strides=y.strides, storage="shared")
+
```
[the remark this answers is not in the document]
Comment (daedalus5) [reply warp/_src/builtins.py:10453]: Can ignore

File: warp/_src/builtins.py [hunk 9]
```diff
@@ -9827,7 +9964,79 @@ def tile_lower_solve_generic_value_func(arg_types, arg_values):
...
+    "tile_lower_solve_inplace",
+    input_types={"L": tile(dtype=Float, shape=Tuple[int, int]), "y": tile(dtype=Float, shape=Tuple[int])},
+    value_func=tile_lower_solve_inplace_generic_value_func,
+    lto_dispatch_func=tile_lower_solve_inplace_generic_lto_dispatch_func,
+    variadic=True,
+    doc="""Solve for z in Lz = y, where L is a lower triangular matrix by overwriting y with z.
+
+    This performs general forward substitution for a lower triangular system inplace.
+
+    Note that computing the adjoint is not yet supported.
```
Comment (daedalus5) [inline warp/_src/builtins.py]: Can remove

File: warp/_src/builtins.py [hunk 10]
```diff
@@ -9960,6 +10153,31 @@ def tile_upper_solve_generic_value_func(arg_types, arg_values):
...
+    "tile_upper_solve_inplace",
+    input_types={"U": tile(dtype=Float, shape=Tuple[int, int]), "z": tile(dtype=Float, shape=Tuple[int])},
+    value_func=tile_upper_solve_inplace_generic_value_func,
+    lto_dispatch_func=tile_upper_solve_inplace_generic_lto_dispatch_func,
+    variadic=True,
+    doc="""Solve for x in Ux = z, where U is an upper triangular matrix by overwriting z with x.
+
+    This performs general back substitution for upper triangular systems inplace.
+
+    Note that computing the adjoint is not yet supported.
```
Comment (daedalus5) [inline warp/_src/builtins.py]: Can remove

File: warp/native/tile.h [hunk 11]
```diff
@@ -3029,27 +3029,30 @@ inline CUDA_CALLABLE void scalar_cholesky(TileA& A, TileL& L)
...
+                L.data(tile_coord(j,k)) = T(0.0);
+            }
         }
     }
 }
 
-// Writes into X
-template <typename TileL, typename TileX, typename TileY>
-inline CUDA_CALLABLE void scalar_cholesky_forward_substitution(TileL& L, TileX& X, TileY& Y)
+// Reads and writes X
```
Comment (daedalus5) [inline warp/native/tile.h]: I'd like to leave open the possibility of writing a backward version of this. Could you instead add a separate `inplace` method?

File: warp/native/tile.h [hunk 12]
```diff
@@ -3121,10 +3124,10 @@ inline CUDA_CALLABLE void scalar_cholesky_back_substitution(TileL& L, TileX& X)
     }
 }
 
-template <typename TileL, typename TileX, typename TileY>
-inline CUDA_CALLABLE void scalar_cholesky_solve(TileL& L, TileX& X, TileY& Y)
+template <typename TileL, typename TileX>
+inline CUDA_CALLABLE void scalar_cholesky_solve(TileL& L, TileX& X)
```
Comment (daedalus5) [inline warp/native/tile.h]: Same comment as above

File: warp/native/tile.h [hunk 13]
```diff
@@ -3319,22 +3322,64 @@ TileL& tile_cholesky(Fwd fun_forward, TileA& A, TileL& L)
...
+    
+    WP_TILE_SYNC();
+
+    // TODO: for batched Cholesky, check all batches
+#if defined(_DEBUG)    
+    if (WP_TILE_THREAD_IDX == 0 && info[0] != 0) {
+        printf("Non-zero status in Cholesky factorization, got %d\n", info[0]);
+    }
+#endif
+
```
Comment (daedalus5) [inline warp/native/tile.h:3647]: Did you find that zeroing the upper triangular part of A was not necessary after cholesky factorization?

File: warp/native/tile.h [hunk 14]
```diff
@@ -3319,22 +3322,64 @@ TileL& tile_cholesky(Fwd fun_forward, TileA& A, TileL& L)
...
+#else
+
+    // TODO: for batched Cholesky, need one info per batch
+    WP_TILE_SHARED int info[1];
+
+    if (WP_TILE_THREAD_IDX == 0) {
+        info[0] = 0;
+    }
+
+    // Call cholesky on L
```
Comment (daedalus5) [inline warp/native/tile.h]: "L" -> "A"

File: warp/native/tile.h [hunk 15]
```diff
@@ -3319,22 +3322,64 @@ TileL& tile_cholesky(Fwd fun_forward, TileA& A, TileL& L)
...
+TileX& tile_cholesky_solve(Fwd fun_forward, TileL& L, TileY& Y, TileX& X)
 {       
     // Copy y to x
 
     X = Y;
 
 #if !defined(__CUDA_ARCH__) || WP_ENABLE_MATHDX == 0
 
-    partitioned_gemm::scalar_cholesky_solve(L, X, Y);
+    partitioned_gemm::scalar_cholesky_solve(L, X);
```
Comment (daedalus5) [inline warp/native/tile.h]: These call changes should be reverted, inplace variants can be used in the new inplace functions

File: warp/native/tile.h (hunk 13 above)
Comment (daedalus5) [reply warp/native/tile.h:3647]: I suppose this is fine if users plan on using Warp solvers. If they only need L however they'll need to be aware that A is not lower triangular. Did you see a major performance difference zeroing vs not zeroing?

**Comment by daedalus5** [comment]
I forgot to mention, it would also be nice to add some additional commentary to the `inplace` method docs. Can mention that they're forward-only, slightly more performant, and require less shared memory.

File: warp/native/tile.h (hunk 13 above)
Comment (RSchwan) [reply warp/native/tile.h:3647]: fig

I was thinking that there should be a performance difference, but turns out that there is no measurable difference (see plot). Hence, I'll also zero the inplace variant.

**Comment by RSchwan** [comment]
All comments have been addressed. Happy to rebase and squash the commits once you've confirmed everything looks good.

File: warp/_src/builtins.py [hunk 16]
```diff
@@ -10125,6 +10140,8 @@ def tile_cholesky_generic_lto_dispatch_func(
         num_threads = options["block_dim"]
         parameter_list = f"({dtype}*, int*)"
         req_smem_bytes = a.type.size * type_size_in_bytes(a.type.dtype)
+        if not inplace:
+            req_smem_bytes *= 2
```
[the remark this answers is not in the document]
Comment (daedalus5) [reply warp/_src/builtins.py:10144]: This isn't an issue. The generated LTO needs the same parameter list regardless of whether the builtin call is inplace or not.

File: warp/native/tile.h [hunk 17]
```diff
@@ -3275,15 +3275,15 @@ inline CUDA_CALLABLE void scalar_matmul(const StorageA& A, const StorageB& B, St
     }
 }
 
-template <typename TileA, typename TileL>
-inline CUDA_CALLABLE void scalar_cholesky(TileA& A, TileL& L)
```
Comment (daedalus5) [inline warp/native/tile.h:3279]: Would you mind doing the same thing here you did with `scalar_cholesky_solve`? ie, keep the existing signature for the inplace cholesky variant, but pass A twice. This will make the (eventual) adjoint implementation easier.

**Comment by daedalus5** [comment]
Thanks, looks good! Please squash and rebase then we'll get the merge process started.

**Comment by daedalus5** [comment]
Hi @RSchwan , would you mind re-basing again and fixing the changelog? Your entry needs to be moved to the next release. It also looks like git is duplicating some existing entries. Sorry this happens when we merge after a release.

**Comment by shi-eric** [comment]
Thanks @RSchwan, this is now merged.
